### DISCUSSION TITLE
Audiotools / pure data

### DISCUSSION THREAD

**Discussion by KatePimbleton** [pschatzmann/arduino-audio-tools#1432] [Q&A] [answered]
Hello,

I wanted to know if there was a way to open a wav, with
Fileloop, then decode it and send it to puredata, in order
to apply effects , before sending it to a i2sStream ???

Cheers

**Comment by pschatzmann** [comment]
I never tried something like this, but I don't see any reason why this should not work...

**Comment by KatePimbleton** [comment]
I gave it a try, 
first I tried the puredata stream example in the arduino audio tools library,
it works as expected, my audiokit produces a tone.

as mentionned in the HVCC wiki in order to receive audio in pure data
you have to create an adc~. So first I created a patch with an "adc~" :
![audioKitTest](https://github.com/pschatzmann/arduino-audio-tools/assets/126875254/2f0d6b03-f7f0-43f4-8537-a638aa64f823)

I compiled the patch then copied the content of "c" in my arduino sketch folder, uploaded the sketch to the audiokit,
after which the audiokit crashes, with the following error message:
"assert failed: void audio_tools::PureDataStream::readWrite(int) PureDataStream.h:87 (eff == sample_size)

Backtrace: 0x40083c85:0x3ffca940 0x4008b97d:0x3ffca960 0x40090aa5:0x3ffca980 0x400d677d:0x3ffcaab0 0x400d6ccf:0x3ffcbaf0 0x400d70af:0x3ffcbb10 0x400d7220:0x3ffcbb50 0x400df27d:0x3ffcbb70"

my end goal is to have 2 Fileloop playing mono wav files, send it to pureDataStream, process them, and
finally output the pureDataStream to AudioKitEs8388V1.

**Comment by KatePimbleton** [comment]
I guess it has something to do with the fact that I don't send anything to the pureDataStream...

**Comment by KatePimbleton** [comment]
I did try again, with one FileLoop:


#include <SPI.h>
#include <SD.h>
#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"
#include "AudioLibs/FileLoop.h"
#include "Heavy_audioKitTest.hpp" 
#include "AudioLibs/PureDataStream.h"


Heavy_audioKitTest pd_test(44100);
PureDataStream pd(pd_test);

const int chipSelect=PIN_AUDIO_KIT_SD_CARD_CS;
AudioBoardStream i2s(AudioKitEs8388V1); // final output of decoded stream
EncodedAudioStream decoder(&pd, new WAVDecoder()); // Decoding stream
// EncodedAudioStream decoder(&i2s, new WAVDecoder()); // Decoding stream

FileLoop loopingFile;

StreamCopy copier(decoder, loopingFile); 
StreamCopy copierPd(i2s,pd); 

void setup(){
  Serial.begin(115200);
  auto config = i2s.defaultConfig(TX_MODE);
  config.sd_active = true;
  i2s.begin(config);

  SD.begin(chipSelect);
  loopingFile.setFile(SD.open("/sean1.wav"));
  decoder.begin();
}

void loop(){
  copier.copy();
 copierPd.copy();
}

I must be wrong, I have no clue how to copy the decoded stream to pd, here I get
this error:

[E] StreamCopy.h : 349 - write  to target has failed after 22 retries! (1024 bytes)

**Comment by pschatzmann** [comment]
This looks right to me. Did you test if the the input chain is working ?

**Reply by pschatzmann** [reply, marked as answer]
Oh, pd.begin() is missing!

**Comment by pschatzmann** [comment]
I just added a [ADC test sketch](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-pd/streams-generator-pd-audiokit) and it is working w/o any problems...

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/2078916/8112e30b-7e24-4f47-908f-9cb45203a748)
